### PR TITLE
feat(caip): update to & from type signatures

### PR DESCRIPTION
**Pull request by 0xApotheosis**
- Adds `chainNamespace` and `chainReference` to CAIP input/output helper type signatures
- By proxy, removes `ChainTypes` and `NetworkTypes` enums from the CAIP package
- Adds type guards and assertion functions to all related types

Implements the CAIP package portions of https://github.com/shapeshift/lib/issues/547 and https://github.com/shapeshift/web/issues/1570 - all changes are isolated to `packages/caip`.

It's a big one due to the number of places these helpers are used (especially tests). Besides some quality of life refactoring along the way and type narrowing for safety, this is about as atomic as we'll get for this piece of work.

Note that this is a breaking change for the CAIP package.

---

This is **Part 1** of a 3-part process.
**Part 2:** update affected `lib` packages
**Part 3:** update `web` to handle breaking changes

### CODE REVIEW

File: packages/caip/src/accountId/accountId.ts [hunk 1]
```diff
@@ -1,39 +1,54 @@
...
-  if (!isChainId(chainId)) {
-    throw new Error(`toAccountId: invalid ChainId ${chainId}`)
-  }
-
-  if (!account) {
-    throw new Error(`toAccountId: account is required`)
-  }
-
-  const [namespace] = chainId.split(':')
+type ToAccountIdArgs =
```
Comment (0xdef1cafe) [inline packages/caip/src/accountId/accountId.ts:7]: 🔥

File: packages/caip/src/adapters/osmosis/utils.ts [hunk 2]
```diff
@@ -45,9 +45,9 @@ export const parseOsmosisData = (data: OsmosisCoin[]) => {
       assetNamespace = 'ibc'
     }
 
-    const chain = ChainTypes.Osmosis
-    const network = NetworkTypes.OSMOSIS_MAINNET
-    const assetId = toAssetId({ chain, network, assetNamespace, assetReference })
+    const chainNamespace = CHAIN_NAMESPACE.Cosmos
```
Comment (0xdef1cafe) [inline packages/caip/src/adapters/osmosis/utils.ts:48]:
```suggestion
    const chainNamespace = CHAIN_NAMESPACE.Osmosis
```

?

File: packages/caip/src/adapters/osmosis/utils.ts [hunk 3]
```diff
@@ -58,8 +58,8 @@ export const parseOsmosisData = (data: OsmosisCoin[]) => {
 
 export const parseData = (d: OsmosisCoin[]) => {
   const osmosisMainnet = toChainId({
-    chain: ChainTypes.Osmosis,
-    network: NetworkTypes.OSMOSIS_MAINNET
+    chainNamespace: CHAIN_NAMESPACE.Cosmos,
```
Comment (0xdef1cafe) [inline packages/caip/src/adapters/osmosis/utils.ts:61]: ?

File: packages/caip/src/typeGuards.ts [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+
+export const isAssetReference = (
+  maybeAssetReference: AssetReference | string
+): maybeAssetReference is AssetReference =>
+  Object.values(ASSET_REFERENCE).some((s) => s === maybeAssetReference)
+
+export const isAssetId = (maybeAssetId: AssetId | string): maybeAssetId is AssetReference => {
+  const matches = parseAssetIdRegExp.exec(maybeAssetId) ?? []
+  return (
+    isChainNamespace(matches[1]) && isChainReference(matches[2]) && isAssetNamespace(matches[3])
```
Comment (0xdef1cafe) [inline packages/caip/src/typeGuards.ts:34]: is there a way we can use named capture groups here? obviously the ordering here is realistically not going to change but would be nice to be able to do `matches['chainReference']`

**Review by 0xdef1cafe — APPROVED** [review]

**Review by cjthompson — APPROVED** [review]

File: packages/caip/src/accountId/accountId.ts [hunk 5]
```diff
@@ -42,24 +57,25 @@ export const fromAccountId: FromAccountId = (accountId) => {
     throw new Error(`fromAccountId: invalid AccountId ${accountId}`)
   }
 
+  const chainNamespace = parts[0]
+  const chainReference = parts[1]
```
Comment (theoboldfrazier) [inline packages/caip/src/accountId/accountId.ts:61]: Given an account id `cosmos:cosmoshub-4/slip44:118`

Won't this give a chainReference of `cosmoshub-4/slip44`, when really the chainReference should be `cosmoshub-4`? `slip44` would be the asset namespace.

**Review by theoboldfrazier — APPROVED** [review]
Looked good overall for me. Pulled down and ran tests.

Is there an associated web PR with this?

**Review by elmutt — APPROVED** [review]

File: packages/caip/src/typeGuards.ts (hunk 4 above)
Comment (0xApotheosis) [reply packages/caip/src/typeGuards.ts]: There is! https://github.com/shapeshift/lib/pull/676/commits/ad56ace6ef4bc1914b2ede3bd18440ca3186f782

**Comment by 0xApotheosis** [comment]
> Is there an associated web PR with this?

Not yet, 🔜

**Review by 0xdef1cafe — APPROVED** [review]

**Review by theoboldfrazier — APPROVED** [review]

**Pull request merged**